prover/src/logup_gkr/mod.rs

```diff
@@ -366,33 +374,40 @@ where
 ///
 /// [1]: https://eprint.iacr.org/2023/1284
 pub fn build_s_column<E: FieldElement>(
-    main_trace: &impl Trace<BaseField = E::BaseField>,
+    trace: &impl Trace<BaseField = E::BaseField>,
     gkr_data: &GkrData<E>,
     evaluator: &impl LogUpGkrEvaluator<BaseField = E::BaseField>,
     lagrange_kernel_col: &[E],
 ) -> Vec<E> {
     let c = gkr_data.compute_batched_claim();
-    let main_segment = main_trace.main_segment();
-    let mean = c / E::from(E::BaseField::from(main_segment.num_rows() as u32));
+    let num_oracles = evaluator.get_oracles().len();
 
-    let mut result = Vec::with_capacity(main_segment.num_rows());
-    let mut last_value = E::ZERO;
-    result.push(last_value);
+    let main_segment = trace.main_segment();
+    let num_cols = main_segment.num_cols();
+    let num_rows = main_segment.num_rows();
+    let mean = c / E::from(E::BaseField::from(num_rows as u32));
 
-    let mut query = vec![E::BaseField::ZERO; evaluator.get_oracles().len()];
-    let mut main_frame = EvaluationFrame::new(main_trace.main_segment().num_cols());
+    let mut deltas = unsafe { uninit_vector(main_segment.num_rows()) };
 
-    for (i, item) in lagrange_kernel_col.iter().enumerate().take(main_segment.num_rows() - 1) {
-        main_trace.read_main_frame(i, &mut main_frame);
+    batch_iter_mut!(&mut deltas[1..], 1024, |batch: &mut [E], batch_offset: usize| {
+        let mut query = vec![E::BaseField::ZERO; num_oracles];
+        let mut main_frame = EvaluationFrame::<E::BaseField>::new(num_cols);
 
-        evaluator.build_query(&main_frame, &mut query);
-        let cur_value = last_value - mean + gkr_data.compute_batched_query(&query) * *item;
+        for (i, v) in batch.iter_mut().enumerate() {
+            trace.read_main_frame(i + batch_offset, &mut main_frame);
 
-        result.push(cur_value);
-        last_value = cur_value;
-    }
-
-    result
+            evaluator.build_query(&main_frame, &mut query);
+            *v = gkr_data.compute_batched_query(&query) * lagrange_kernel_col[i + batch_offset]
+                - mean;
+        }
+    });
+
+    // note that `deltas[0]` is set `0` and thus `deltas` satisfies the conditions for invoking
+    // the function
+    let mut cumulative_sum = deltas;
+    cumulative_sum[0] = E::ZERO;
```

Comment:
nit: can we set `deltas[0] - E::ZERO` on line 391 so that it's clear before we get to the `cumulative_sum` part that all elements on `deltas` were properly initialized?

Reply:
Makes sense, updated